main.go

```diff
@@ -1,188 +1,50 @@
-		http.Error(w, "post not found", http.StatusNotFound)
-		return
-	}
-
-	post := s.posts[i]
-	resp, err := json.Marshal(post)
-	if err != nil {
-		http.Error(w, "internal server error", http.StatusInternalServerError)
-		return
-	}
-
-	if err := json.NewEncoder(w).Encode(resp); err != nil {
-		http.Error(w, "failed to write response", http.StatusInternalServerError)
-	}
-}
-
-func (s *store) updatePost(w http.ResponseWriter, r *http.Request) {
-	id, err := strconv.Atoi(chi.URLParam(r, "id"))
-	if err != nil {
-		http.Error(w, "invalid id", http.StatusBadRequest)
-		return
-	}
-
-	decoder := json.NewDecoder(r.Body)
-	var p post
-	if err := decoder.Decode(&p); err != nil {
-		http.Error(w, "invalid request body", http.StatusBadRequest)
-	}
-
-	s.m.Lock()
-	defer s.m.Unlock()
-
-	i, ok := s.find(id)
-	if !ok {
-		http.Error(w, "post not found", http.StatusNotFound)
-		return
-	}
-
-	updated := s.posts[i]
-	updated.Body = p.Body
-	updated.Title = p.Title
-	s.posts[i] = updated
-
-	resp, err := json.Marshal(updated)
-	if err != nil {
-		http.Error(w, "internal server error", http.StatusInternalServerError)
-		return
-	}
-
-	if err := json.NewEncoder(w).Encode(resp); err != nil {
-		http.Error(w, "failed to write response", http.StatusInternalServerError)
-	}
-}
-
-func (s *store) createPost(w http.ResponseWriter, r *http.Request) {
-	decoder := json.NewDecoder(r.Body)
-	var p post
-	if err := decoder.Decode(&p); err != nil {
-		http.Error(w, "invalid request body", http.StatusBadRequest)
-		return
-	}
-
-	id := 1
-	s.m.Lock()
-	count := len(s.posts)
-	if count > 0 {
-		id = s.posts[count-1].ID + 1
-	}
-
-	p.ID = id
-
-	s.posts = append(s.posts, p)
-	s.m.Unlock()
-
-	resp, err := json.Marshal(p)
-	if err != nil {
-		http.Error(w, "internal server error", http.StatusInternalServerError)
-		return
-	}
-
-	if err := json.NewEncoder(w).Encode(resp); err != nil {
-		http.Error(w, "failed to write response", http.StatusInternalServerError)
-	}
-}
+func main() {
+	dbURL := os.Getenv("DATABASE_URL")
 
-func (s *store) deletePost(w http.ResponseWriter, r *http.Request) {
-	id, err := strconv.Atoi(chi.URLParam(r, "id"))
+	m, err := migrate.New("file://migrations", dbURL)
 	if err != nil {
-		http.Error(w, "invalid id", http.StatusBadRequest)
-		return
+		log.Fatalln(err)
 	}
 
-	s.m.Lock()
-	defer s.m.Unlock()
-
-	i, ok := s.find(id)
-	if !ok {
-		http.Error(w, "post not found", http.StatusNotFound)
-		return
+	if err := m.Up(); err != nil {
+		if err == migrate.ErrNoChange {
+			log.Println("MIGRATION: no database change")
+		} else {
+			log.Fatalln(err)
+		}
 	}
 
-	s.posts[i] = s.posts[len(s.posts)-1]
-	s.posts[len(s.posts)-1] = post{}
-	s.posts = s.posts[:len(s.posts)-1]
-
-	fmt.Fprintf(w, "Deleted")
-}
-
-func main() {
-	s := &store{
-		[]post{{
-			ID:    1,
-			Body:  "You",
-			Title: "YOusdkfkd",
-		}},
-		&sync.RWMutex{},
-	}
+	db, teardown := db.New(dbURL)
+	defer func() {
+		if err := teardown(); err != nil {
+			log.Fatalln(err)
+		}
+	}()
```

Comment:
I don't know if this is really necessary because it is in the main function. I'm just doing it this way because calling `teardown` (`postgres.Close`) returns an error.

Reply:
This is correct, note that `log.Fatalln` calls `os.Exit(1)` which will not run deferred functions, so if you want to do this in main, you shouldn't use `log.Fatalln`. I would suggest putting this in a separate function that you call from main, so you can use `log.Fatalln` in `main`, but you can otherwise return errors. Seehttps://pace.dev/blog/2020/02/12/why-you-shouldnt-use-func-main-in-golang-by-mat-ryer.html for an example.